std-bits/base/src/main/java/org/enso/base/enso_cloud/TransientHTTPResponseCache.java

```diff
@@ -0,0 +1,428 @@
+    assert maxTotalCacheSizeOverrideTestOnly_ <= MAX_TOTAL_CACHE_SIZE;
+    maxTotalCacheSizeOverrideTestOnly = Optional.of(maxTotalCacheSizeOverrideTestOnly_);
+  }
+
+  public static void clearMaxTotalCacheSizeOverrideTestOnly() {
+    maxTotalCacheSizeOverrideTestOnly = Optional.empty();
+  }
+
+  private static ZonedDateTime getNow() {
+    // The 'nowOverrideTestOnly ' field is used for testing TTL expiration logic.
+    if (nowOverrideTestOnly != null) {
+      return nowOverrideTestOnly;
+    } else {
+      return ZonedDateTime.now();
+    }
+  }
+
+  /**
+   * We define the TTL as the amount of time that the response should be considered fresh.
+   *
+   * <p>Define t0 as the time at which the content was generated on the origin server.
+   *
+   * <p>Define t1 as the time at which the current request was handled, either by the origin server
+   * or an intervening cache.
+   *
+   * <p>The 'Age' header, if present is (t1 - t0).
+   *
+   * <p>The 'max-age' value in the 'Cache-Control' header, if present, is the origin server's
+   * definition of how long the response should be considered fresh.
+   *
+   * <p>If 'max-age' and 'Age' are both present, we set TTL = max-age - Age. If only 'max-age' is
+   * present, we set TTL = max-age. If neither are present, we use a default.
+   */
+  private static int calculateTTL(HttpHeaders headers) {
+    Integer maxAge = getMaxAge(headers);
+    if (maxAge == null) {
+      return DEFAULT_TTL_SECONDS;
+    } else {
+      int age = headers.firstValue("age").map(Integer::parseInt).orElse(0);
+      return maxAge - age;
+    }
+  }
+
+  private static Integer getMaxAge(HttpHeaders headers) {
+    var cacheControlMaybe = headers.firstValue("cache-control");
+    Integer maxAge = null;
+    if (cacheControlMaybe.isPresent()) {
+      var cacheControl = cacheControlMaybe.get();
+      var cacheControlEntries = cacheControl.split(",");
+      for (var entry : cacheControlEntries) {
+        if (entry.trim().toLowerCase().startsWith("max-age")) {
+          var maxAgeBinding = entry.split("=");
+          if (maxAgeBinding.length > 1) {
+            maxAge = Integer.parseInt(maxAgeBinding[1]);
+          }
+          break;
+        }
+      }
+    }
+    return maxAge;
+  }
+
+  /**
+   * Sorts the header by header name, so we don't depend on header order. Multiple-valued headers
+   * might hash differently, but it's a rare case.
+   */
+  private static String makeHashKey(URI resolvedURI, List<Pair<String, String>> resolvedHeaders) {
+    try {
+      MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
+      messageDigest.update(resolvedURI.toString().getBytes());
+
+      var sortedHeaders = resolvedHeaders.stream().sorted(headerNameComparator).toList();
+      for (Pair<String, String> resolvedHeader : sortedHeaders) {
+        messageDigest.update(resolvedHeader.getLeft().getBytes());
+        messageDigest.update(resolvedHeader.getRight().getBytes());
+      }
```

Comment:
Could we use just use Java's string hash?
Any reason to use SHA-256 over faster MD5?

Reply:
Do you mean concatenate the uri and headers and use String.hashCode()? I used this for better security, but can definitely switch to MD5.

Reply:
SHA256 is heavier to compute than a simple MD5 - not sure makes too much difference on security as reversing an MD5 is pretty hard anyway, but as this is on headers and uri its not a real issue.

The Java string hash was just couldn't we just make a big string and hash it - but I prefer yours.

Reply:
> used this for better security

Security of what? I mean:
- there is no check whether the first download of the HTTP resource is trustworthy, right?
- e.g. the system downloads everything and just puts a stamp on it
- if someone can get access to a computer (user account) then there is no security anyway
- so: what kind of attack this hash is shielding the user against?

Reply:
Synthesizing a URL that would hash to an existing resource in the user's cache that they hadn't intended to load. It's an extremely unlikely case, but I thought of it just as the routine use of non-trivial hashes for resource naming, and the cost is once per HTTP request. I'm fine using `hashCode` for this.